File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/docker/DockerComposeConfigurationSettingsApplier.java
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.docker;

import static java.util.stream.Collectors.toList;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerComposeConfigurationOptionsSettings;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerComposeConfigurationSettings;
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerEnvironmentVariable;
import com.intellij.docker.DockerDeploymentConfiguration;
import com.intellij.docker.DockerRunConfigurationCreator;
import com.intellij.docker.agent.settings.DockerEnvVarImpl;
import com.intellij.docker.deploymentSource.DockerComposeDeploymentSourceType;
import com.intellij.execution.RunManager;
import com.intellij.execution.RunnerAndConfigurationSettings;
import java.nio.file.Path;
import java.util.Optional;
import javax.inject.Inject;

public class DockerComposeConfigurationSettingsApplier implements SettingsApplier<DockerComposeConfigurationSettings> {

  private final DockerRunConfigurationCreator dockerRunConfigurationCreator;
  private final RunManager runManager;

  @Inject
  public DockerComposeConfigurationSettingsApplier(DockerRunConfigurationCreator dockerRunConfigurationCreator,
                                                   RunManager runManager) {
    this.dockerRunConfigurationCreator = dockerRunConfigurationCreator;
    this.runManager = runManager;
  }

  @Override
  public void apply(DockerComposeConfigurationSettings settings) {
    final DockerComposeDeploymentSourceType dockerComposeDeploymentSourceType = DockerComposeDeploymentSourceType.getInstance();
    final DockerDeploymentConfiguration dockerDeploymentConfiguration = new DockerDeploymentConfiguration();

    settings.services()
        .ifPresent(dockerDeploymentConfiguration::setServices);

    settings.composeFiles()
        .map(composeFiles -> composeFiles.stream()
            .map(Path::toAbsolutePath)
            .map(Path::toString)
            .collect(toList()))
        .flatMap(composeFiles -> composeFiles.isEmpty()
            ? Optional.empty()
            : Optional.of(composeFiles))
        .ifPresent(composeFiles -> {
            dockerDeploymentConfiguration.setSourceFilePath(composeFiles.get(0));
            dockerDeploymentConfiguration.setSecondarySourceFiles(composeFiles.subList(1, composeFiles.size()));
        });

    settings.environmentVariables().ifPresent(variables -> dockerDeploymentConfiguration.setEnvVars(
        variables.stream()
            .map(this::toDockerEnvVarImpl)
            .collect(toList())));

    settings.options()
        .flatMap(DockerComposeConfigurationOptionsSettings::buildForceBuildImages)
        .ifPresent(build -> dockerComposeDeploymentSourceType.applyForceBuild(dockerDeploymentConfiguration, build));

    final RunnerAndConfigurationSettings runnerAndConfigurationSettings = dockerRunConfigurationCreator.createConfiguration(
        dockerComposeDeploymentSourceType.getSingletonSource(),
        dockerDeploymentConfiguration,
        null);
    runnerAndConfigurationSettings.setName(settings.name());
    runManager.addConfiguration(runnerAndConfigurationSettings);
  }

  private DockerEnvVarImpl toDockerEnvVarImpl(DockerEnvironmentVariable dockerEnvironmentVariable) {
    DockerEnvVarImpl dockerEnvVarImpl = new DockerEnvVarImpl();
    dockerEnvironmentVariable.name().ifPresent(dockerEnvVarImpl::setName);
    dockerEnvironmentVariable.value().ifPresent(dockerEnvVarImpl::setValue);
    return dockerEnvVarImpl;
  }
}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/docker/DockerComposeConfigurationSettingsApplier.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.docker

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerComposeConfigurationSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.docker.DockerEnvironmentVariable
import com.intellij.docker.DockerDeploymentConfiguration
import com.intellij.docker.DockerRunConfigurationCreator
import com.intellij.docker.agent.settings.DockerEnvVarImpl
import com.intellij.docker.deploymentSource.DockerComposeDeploymentSourceType
import com.intellij.execution.RunManager
import java.util.Optional
import javax.inject.Inject

class DockerComposeConfigurationSettingsApplier @Inject
constructor(private val dockerRunConfigurationCreator: DockerRunConfigurationCreator,
            private val runManager: RunManager) : SettingsApplier<DockerComposeConfigurationSettings> {

    override fun apply(settings: DockerComposeConfigurationSettings) {
        val dockerComposeDeploymentSourceType = DockerComposeDeploymentSourceType.getInstance()
        val dockerDeploymentConfiguration = DockerDeploymentConfiguration()

        settings.services?.let { dockerDeploymentConfiguration.services = it }

        settings.composeFiles
            ?.let { composeFiles ->
                composeFiles
                    .map { it.toAbsolutePath() }
                    .map { it.toString() }
            }
            ?.let { composeFiles -> if (composeFiles.isNotEmpty()) composeFiles else null }
            ?.let { composeFiles ->
                dockerDeploymentConfiguration.sourceFilePath = composeFiles[0]
                dockerDeploymentConfiguration.secondarySourceFiles = composeFiles.subList(1, composeFiles.size)
            }

        settings.environmentVariables?.let { variables -> dockerDeploymentConfiguration.envVars = variables.map { toDockerEnvVarImpl(it) } }

        settings.options?.buildForceBuildImages?.let { dockerComposeDeploymentSourceType.applyForceBuild(dockerDeploymentConfiguration, it) }

        val runnerAndConfigurationSettings = dockerRunConfigurationCreator.createConfiguration(
            dockerComposeDeploymentSourceType.singletonSource,
            dockerDeploymentConfiguration,
            null)
        runnerAndConfigurationSettings.name = settings.name
        runManager.addConfiguration(runnerAndConfigurationSettings)
    }

    private fun toDockerEnvVarImpl(dockerEnvironmentVariable: DockerEnvironmentVariable): DockerEnvVarImpl {
        val dockerEnvVarImpl = DockerEnvVarImpl()
        dockerEnvironmentVariable.name.let { dockerEnvVarImpl.name = it }
        dockerEnvironmentVariable.value?.let { dockerEnvVarImpl.value = it }
        return dockerEnvVarImpl
    }
}
